### ISSUE TITLE
useDataProvider returns invalid data for lists after checkError rejects

### ISSUE THREAD

**Issue by hlubek** [marmelab/react-admin#7911]
**What you were expecting:**

* A user is logged out after an operation failed due to invalid authentication (`checkError` rejects) and can log in again without errors

**What happened instead:**

<img width="1208" alt="react-admin-issue" src="https://user-images.githubusercontent.com/33351/176513020-1624d5ce-1b3b-42a5-8d0d-d1c79ebcd40b.png">

* In this case `useDataProvider` calls `logoutIfAccessDenied` and returns `{ data: {} }` if the user was logged out
* The value `{}` is not valid for a `Datagrid` so the page shows an error and there will be lots of errors in the console (see screenshot)

**Steps to reproduce:**

* Add an auth provider
* Add any resource with a list
* Create a scenario where `checkAuth` still returns `true`, but `checkErr` will fail
* Open the root path `/`
* Log in

**Environment**

* React-admin version: 4.2.0

**Comment by septentrion-730n** [comment]
Thanks for the report. Could you provide a codesandbox demo for reproduction using the [template](https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple) ?

**Comment by hlubek** [comment]
> Thanks for the report. Could you provide a codesandbox demo for reproduction using the [template](https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple) ?

Yes, it suffices to add an authentication check to the data provider:

https://119yp6.sse.codesandbox.io/

Just open the root, log in and the error `data.map` is not a function should occur.

**Comment by joobacca** [comment]
Any updates on this? I have the same issue and had to turn off the react-query cache as a workaround.
```es6

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      ...someOtherOptions,
      cacheTime: 0,
      staleTime: 0,
    },
  },
});
```

Is there a better way to handle this problem?

**Comment by fzaninotto** [comment]
Fixed by #7991
